README.md

[:heart: sponsor](https://github.com/sponsors/rbellens)

# x509

Dart library for parsing and working with X.509 certificates.


## Usage

A simple usage example:
```dart

import 'package:x509/x509.dart';
import 'dart:io';

void main() {
  var cert = parsePem(File('cert.pem').readAsStringSync());
      
  print(cert);
}
```
## Features and bugs

Please file feature requests and bugs at the [issue tracker][tracker].

[tracker]: https://github.com/appsup-dart/x509/issues



## Sponsor

Creating and maintaining this package takes a lot of time. If you like the result, please consider to [:heart: sponsor](https://github.com/sponsors/rbellens). 
With your support, I will be able to further improve and support this project.
Also, check out my other dart packages at [pub.dev](https://pub.dev/packages?q=publisher%3Aappsup.be).




CHANGELOG.md
## 0.2.4+2

 - **DOCS**: fix code sample (pull request [#29](https://github.com/appsup-dart/x509/issues/29) from bivens-dev/patch-1). ([14a3a777](https://github.com/appsup-dart/x509/commit/14a3a777afae604e293ee18eb34629c3ef875862))

## 0.2.4+1

 - **FIX**: remove unnecessary null checks. ([fb37bbed](https://github.com/appsup-dart/x509/commit/fb37bbeddd1f76dc027c9b189ea2e016383971c5))

## 0.2.4

 - **FIX**: parse the 1.3.6.1.4.1.11129.2.4.2 OID used for the SCT list extension. ([9349b174](https://github.com/appsup-dart/x509/commit/9349b174fbce45242bbdef154bc96bde5b20e781))
 - **FEAT**: allow parsing of unknown extensions if they are non-critical (pull request [#22](https://github.com/appsup-dart/x509/issues/22) from sroddy). ([74b98a9c](https://github.com/appsup-dart/x509/commit/74b98a9c34884ec995646ac5716c81aec807b488))

## 0.2.3+1

 - **FIX**: null-aware warnings (pull request [#25](https://github.com/appsup-dart/x509/issues/25) from faithoflifedev). ([d3f7fcb9](https://github.com/appsup-dart/x509/commit/d3f7fcb9956beefc6f41e67832a824304d09210b))

## 0.2.3

 - **FIX**: parsing certificates containing GeneralizedTime (pull request [#21](https://github.com/appsup-dart/x509/issues/21) from adamgillmore). ([850c55fb](https://github.com/appsup-dart/x509/commit/850c55fb60f4ebf705f5c3f5481635e5a4f498a3))
 - **FIX**: wrong cast in ObjectIdentifier.toAsn1() (pull request [#14](https://github.com/appsup-dart/x509/issues/14) from NicolaVerbeeck). ([95e838ca](https://github.com/appsup-dart/x509/commit/95e838ca08b5d049fb5bf6a29eecfd6486e89dcb))
 - **FEAT**: ProxyCertInfo extension. ([17df803c](https://github.com/appsup-dart/x509/commit/17df803c9423c2d6329abfb8880353f46a78d145))
 - **FEAT**: NameConstraints extension. ([db387ef2](https://github.com/appsup-dart/x509/commit/db387ef29b76e7c41d59ce267f2a6365b016c5b0))
 - **FEAT**: QCStatements extension (pull request [#15](https://github.com/appsup-dart/x509/issues/15) from jeroentrappers). ([a5bbd73c](https://github.com/appsup-dart/x509/commit/a5bbd73ce9b1c3a30d063e4a4eed2d64837195bb))
 - **FEAT**: PrivateKeyUsagePeriod extension (pull request [#15](https://github.com/appsup-dart/x509/issues/15) from jeroentrappers). ([27cc8f50](https://github.com/appsup-dart/x509/commit/27cc8f5062665a0d6c873db529ab8bf6981c7556))
 - **FEAT**: support unknown policyidentifier (pull request [#12](https://github.com/appsup-dart/x509/issues/12) from nakajo2011). ([bb76649a](https://github.com/appsup-dart/x509/commit/bb76649a4abfc44a201cea4a66a6d2e4fd2a4187))
 - **DOCS**: add funding info. ([0b2a91ac](https://github.com/appsup-dart/x509/commit/0b2a91ac57acb7a632396fc410bedd8ba6df0aff))

## 0.2.2

- Compatible with version `0.3.0` of `crypto_keys`

## 0.2.1

- Parse pem certificates

## 0.2.0

- Migrate null safety

## 0.1.4

- Bump `asn1lib` to 0.8.1.
- Support GeneralNames for using by Issuer/Subject Alternative Name

## 0.1.3

- Support for CertificatePolicies, CrlDistributionPoints and AuthorityInformationAccess extensions

## 0.1.2

- Support for EC keys

## 0.1.1

- Bugfix parsing strings stored as ASN1UTF8String

## 0.1.0

- Initial version


test/x509.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:test/test.dart';
import 'package:x509/x509.dart';
import 'dart:io';
import 'package:asn1lib/asn1lib.dart';

void main() {
  group('rsa', () {
    test('parse key', () {
      var pem = File('test/files/rsa.key').readAsStringSync();
      KeyPair keyPair = parsePem(pem).single;
      var privateKey = keyPair.privateKey as RsaPrivateKey;
      var publicKey = keyPair.publicKey as RsaPublicKey;
      expect(privateKey.firstPrimeFactor, isNotNull);
      expect(privateKey.secondPrimeFactor, isNotNull);
      expect(privateKey.privateExponent, isNotNull);
      expect(publicKey.exponent, isNotNull);
    });
  });

  group('ec', () {
    test('parse ec 256 public key', () {
      var pem = File('test/files/ec256.public.key').readAsStringSync();
      SubjectPublicKeyInfo keyInfo = parsePem(pem).single;
      var key = keyInfo.subjectPublicKey as EcPublicKey;
      expect(
          key.xCoordinate,
          BigInt.parse(
              '51818492006749570326812946343209411041700944121241362625086788703158476893928'));
      expect(
          key.yCoordinate,
          BigInt.parse(
              '33595934851494958356084148292611983061011944871851202520631261652578803383469'));
    });
    test('parse ec 256k public key', () {
      var pem = File('test/files/ec256k.pub.key').readAsStringSync();
      SubjectPublicKeyInfo keyInfo = parsePem(pem).single;
      var key = keyInfo.subjectPublicKey as EcPublicKey;
      expect(key.curve, curves.p256k);
    });
    test('parse ec 256 private key', () {
      var pem = File('test/files/ec256.private.key').readAsStringSync();
      KeyPair keyPair = parsePem(pem).single;
      var key = keyPair.privateKey as EcPrivateKey;

      expect(
          key.eccPrivateKey,
          BigInt.parse(
              '115735426896566426443735562247805583061594925902272203105205578087471800291450'));
    });
    test('parse ec 384 public key', () {
      var pem = File('test/files/ec384.public.key').readAsStringSync();
      SubjectPublicKeyInfo keyInfo = parsePem(pem).single;
      var key = keyInfo.subjectPublicKey as EcPublicKey;
      expect(
          key.xCoordinate,
          BigInt.parse(
              '20545964214668137657626333380687804621770301889137394027403195865297761228836360480027285257779141677515403916001231'));
      expect(
          key.yCoordinate,
          BigInt.parse(
              '13068848895562259854033738524358881457347119109053595396837669867699203318147516158760410558095930543191355315446383'));
    });
    test('parse ec 384 private key', () {
      var pem = File('test/files/ec384.private.key').readAsStringSync();
      KeyPair keyPair = parsePem(pem).single;
      var key = keyPair.privateKey as EcPrivateKey;

      expect(
          key.eccPrivateKey,
          BigInt.parse(
              '30758094300428071899891161382675739865625031661974292037243998715665065103389256758945325690612942812921541455724491'));
    });
    test('parse ec 521 public key', () {
      var pem = File('test/files/ec521.public.key').readAsStringSync();
      SubjectPublicKeyInfo keyInfo = parsePem(pem).single;
      var key = keyInfo.subjectPublicKey as EcPublicKey;
      expect(
          key.xCoordinate,
          BigInt.parse(
              '6558566456959953544109522959384633002634366184193672267866407124696200040032063394775499664830638630438428532794662648623689740875293641365317574204038644132'));
      expect(
          key.yCoordinate,
          BigInt.parse(
              '705914061082973601048865942513844186912223650952616397119610620188911564288314145208762412315826061109317770515164005156360031161563418113875601542699600118'));
    });
    test('parse ec 521 private key', () {
      var pem = File('test/files/ec521.private.key').readAsStringSync();
      KeyPair keyPair = parsePem(pem).single;
      var key = keyPair.privateKey as EcPrivateKey;

      expect(
          key.eccPrivateKey,
          BigInt.parse(
              '5341829702302574813496892344628933729576493483297373613204193688404465422472930583369539336694834830511678939023627363969939187661870508700291259319376559490'));
    });
    test('parse ec 256 key pair', () {
      var pem = File('test/files/ec256.key').readAsStringSync();

      KeyPair keyPair = parsePem(pem).single;

      var privateKey = keyPair.privateKey as EcPrivateKey;
      var publicKey = keyPair.publicKey as EcPublicKey;
      expect(privateKey.eccPrivateKey, isNotNull);
      expect(privateKey.curve, curves.p256);
      expect(publicKey.curve, curves.p256);
      expect(publicKey.xCoordinate, isNotNull);
      expect(publicKey.yCoordinate, isNotNull);
      var signature = privateKey
          .createSigner(algorithms.signing.ecdsa.sha256)
          .sign('hello world'.codeUnits);

      var verified = publicKey
          .createVerifier(algorithms.signing.ecdsa.sha256)
          .verify(Uint8List.fromList('hello world'.codeUnits), signature);

      expect(verified, isTrue);
    });
    test('parse ec 256k key pair', () {
      var pem = File('test/files/ec256k.key').readAsStringSync();

      KeyPair keyPair = parsePem(pem).single;

      var privateKey = keyPair.privateKey as EcPrivateKey;
      var publicKey = keyPair.publicKey as EcPublicKey;
      expect(privateKey.eccPrivateKey, isNotNull);
      expect(privateKey.curve, curves.p256k);
      expect(publicKey.curve, curves.p256k);
      expect(publicKey.xCoordinate, isNotNull);
      expect(publicKey.yCoordinate, isNotNull);
      var signature = privateKey
          .createSigner(algorithms.signing.ecdsa.sha256)
          .sign('hello world'.codeUnits);

      var verified = publicKey
          .createVerifier(algorithms.signing.ecdsa.sha256)
          .verify(Uint8List.fromList('hello world'.codeUnits), signature);

      expect(verified, isTrue);
    });
    test('parse ec 384 key pair', () {
      var pem = File('test/files/ec384.key').readAsStringSync();

      KeyPair keyPair = parsePem(pem).single;

      var privateKey = keyPair.privateKey as EcPrivateKey;
      var publicKey = keyPair.publicKey as EcPublicKey;
      expect(privateKey.eccPrivateKey, isNotNull);
      expect(privateKey.curve, curves.p384);
      expect(publicKey.curve, curves.p384);
      expect(publicKey.xCoordinate, isNotNull);
      expect(publicKey.yCoordinate, isNotNull);
      var signature = privateKey
          .createSigner(algorithms.signing.ecdsa.sha384)
          .sign('hello world'.codeUnits);

      var verified = publicKey
          .createVerifier(algorithms.signing.ecdsa.sha384)
          .verify(Uint8List.fromList('hello world'.codeUnits), signature);

      expect(verified, isTrue);
    });
  });

  group('csr', () {
    test('parse csr', () {
      var pem = File('test/files/csr.pem').readAsStringSync();
      parsePem(pem).single as CertificationRequest;
    });
  });

  group('rfc5280', () {
    test('RSA Self-Signed Certificate', () {
      var f = File('test/resources/rfc5280_cert1.cer');

      var bytes = f.readAsBytesSync();

      var c = X509Certificate.fromAsn1(
          ASN1Parser(bytes).nextObject() as ASN1Sequence);
      expect(c, isA<X509Certificate>());
    });
    test('Apple certificate for server-based Game Center verification', () {
      var f = File('test/resources/3rd-party-auth-prod-19824d.cer');

      var bytes = f.readAsBytesSync();
      var c = X509Certificate.fromAsn1(
          ASN1Parser(bytes).nextObject() as ASN1Sequence);
      expect(c, isA<X509Certificate>());
    });
  });

  group('v3 extension', () {
    var generalNameEncodeBytes = [
      48,
      19,
      130,
      17,
      119,
      119,
      119,
      46,
      99,
      104,
      97,
      105,
      110,
      116,
      111,
      112,
      101,
      46,
      99,
      111,
      109
    ];
    test('subject alternative name(=GeneralNames)', () {
      var extension =
          ASN1Sequence.fromBytes(Uint8List.fromList(generalNameEncodeBytes));
      var oid = ObjectIdentifier([2, 5, 29, 17]);
      var c = ExtensionValue.fromAsn1(extension, oid);
      expect(c, isA<GeneralNames>());
    });
    test('can parse DNS of subjectAltName', () {
      var extension =
          ASN1Sequence.fromBytes(Uint8List.fromList(generalNameEncodeBytes));
      var oid = ObjectIdentifier([2, 5, 29, 17]);
      var c = ExtensionValue.fromAsn1(extension, oid) as GeneralNames;
      expect(c.names[0].toString(), 'DNS:www.chaintope.com');
    });
  });

  group('keys from auth services', () {
    test('https://login.microsoftonline.com/consumers/discovery/v2.0/keys', () {
      var f = json
          .decode(File('test/files/microsoft_keys.json').readAsStringSync());

      for (var k in f['keys'] as List) {
        for (var v in k['x5c'] as List) {
          var bytes = base64.decode(v);
          var p = ASN1Parser(bytes);
          var o = p.nextObject();
          if (o is! ASN1Sequence) {
            throw FormatException('Expected SEQUENCE, got ${o.runtimeType}');
          }
          var s = o;
          print(X509Certificate.fromAsn1(s));
        }
      }
    });
  });
}


test/objectidentifier_test.dart
import 'package:asn1lib/asn1lib.dart';
import 'package:test/test.dart';
import 'package:x509/x509.dart';

void main() {
  group('equality', (){
    test('equals',(){
      expect(ObjectIdentifier([2, 5, 4, 3]) == ObjectIdentifier([2, 5, 4, 3]), true);
    });
    test('not equals last node different',(){
      expect(ObjectIdentifier([2, 5, 4, 3]) != ObjectIdentifier([2, 5, 4, 4]), true);
    });
    test('not equals last node same',(){
      expect(ObjectIdentifier([2, 5, 4, 3]) != ObjectIdentifier([2, 5, 8, 3]), true);
    });
  });
  
  group('name', () {
    test('return correct name', () {
      var oid = ObjectIdentifier([2, 5, 4, 3]);
      expect(oid.name, 'commonName');
    });

    test('throw UnknownOIDNameError when unknown oid', () {
      var oid = ObjectIdentifier([1, 2, 3, 4, 5, 6]);
      expect(() => oid.name, throwsA(TypeMatcher<UnknownOIDNameError>()));
    });
  });
  
  group('asn.1 conversion', (){
    test('convert to asn.1', (){
      var oidOriginal = ObjectIdentifier([2, 5, 4, 3]);
      var asn1Original = oidOriginal.toAsn1();
      var bytes = asn1Original.encodedBytes;
      var asn1Restored = ASN1ObjectIdentifier.fromBytes(bytes);
      var oidRestored = ObjectIdentifier.fromAsn1(asn1Restored);
      expect(oidOriginal == oidRestored, true);
    });
    test('convert from asn.1', (){
      var oidOriginal = ObjectIdentifier([2, 5, 4, 3]);
      var asn1Original = ASN1ObjectIdentifier([2, 5, 4, 3]);
      var oidFromAsn1 = ObjectIdentifier.fromAsn1(asn1Original);
      expect(oidOriginal == oidFromAsn1, true);
    });
  });
}

lib/src/request.dart
part of x509;

/// https://tools.ietf.org/html/rfc2986
class CertificationRequest {
  final CertificationRequestInfo certificationRequestInfo;
  final AlgorithmIdentifier signatureAlgorithm;
  final Uint8List signature;

  CertificationRequest(
      this.certificationRequestInfo, this.signatureAlgorithm, this.signature);

  /// CertificationRequest ::= SEQUENCE {
  ///   certificationRequestInfo CertificationRequestInfo,
  ///   signatureAlgorithm AlgorithmIdentifier{{ SignatureAlgorithms }},
  ///   signature          BIT STRING
  /// }
  factory CertificationRequest.fromAsn1(ASN1Sequence sequence) {
    final algorithm = AlgorithmIdentifier.fromAsn1(sequence.elements[1] as ASN1Sequence);
    return CertificationRequest(
        CertificationRequestInfo.fromAsn1(sequence.elements[0] as ASN1Sequence),
        algorithm,
        (sequence.elements[2] as ASN1BitString).contentBytes());
  }
}

class CertificationRequestInfo {
  final int? version;
  final Name subject;
  final SubjectPublicKeyInfo subjectPublicKeyInfo;
  final Map<String, List<dynamic>>? attributes;

  CertificationRequestInfo(
      this.version, this.subject, this.subjectPublicKeyInfo, this.attributes);

  /// CertificationRequestInfo ::= SEQUENCE {
  ///   version       INTEGER { v1(0) } (v1,...),
  ///   subject       Name,
  ///   subjectPKInfo SubjectPublicKeyInfo{{ PKInfoAlgorithms }},
  ///   attributes    [0] Attributes{{ CRIAttributes }}
  /// }
  factory CertificationRequestInfo.fromAsn1(ASN1Sequence sequence) {
    return CertificationRequestInfo(
        toDart(sequence.elements[0]).toInt() + 1,
        Name.fromAsn1(sequence.elements[1] as ASN1Sequence),
        SubjectPublicKeyInfo.fromAsn1(sequence.elements[2] as ASN1Sequence),
        null /*TODO*/);
  }
}


lib/src/objectidentifier.dart
part of x509;

class ObjectIdentifier {
  final List<int> nodes;

  const ObjectIdentifier(this.nodes);

  ObjectIdentifier? get parent => nodes.length > 1
      ? ObjectIdentifier(nodes.take(nodes.length - 1).toList())
      : null;

  factory ObjectIdentifier.fromAsn1(ASN1ObjectIdentifier id) {
    // ASN1ObjectIdentifier parameter 'oi' is the same as ObjectIdentifier parameter nodes
    return ObjectIdentifier(id.oi);
  }

  ASN1ObjectIdentifier toAsn1() {
    // ObjectIdentifier parameter 'oi' is the same as ASN1ObjectIdentifier parameter nodes
    return ASN1ObjectIdentifier(nodes);
  }

  @override
  int get hashCode => hashObjects(nodes);

  @override
  bool operator ==(dynamic other) =>
      other is ObjectIdentifier && listsEqual(nodes, other.nodes);

  String get name {
    try {
      dynamic tree = _tree;
      for (var n in nodes) {
        tree = tree[n];
      }
      if (tree is Map) return tree[null];
      return tree;
    } catch (e) {
      throw UnknownOIDNameError(
          'Unable to get name of ObjectIdentifier with nodes $nodes');
    }
  }

  @override
  String toString() {
    try {
      return name;
    } on UnknownOIDNameError {
      return nodes.join('.');
    }
  }

  static const _tree = {
    0: {
      null: 'itu-t',
      9: {
        null: 'data',
        2342: {
          null: 'pss',
          19200300: {
            null: 'ucl',
            100: {
              null: 'pilot',
              1: {
                null: 'pilotAttributeType',
                25: {null: 'domainComponent'}
              }
            }
          }
        }
      },
      4: {
        null: 'identified-organization',
        0: {
          null: 'etsi',
          1862: {
            null: 'qc-profile',
            0: {
              null: 'id-mod',
              2: 'id-mod-qc-profile-2',
            },
            1: {
              null: 'qcs',
              1: 'qcs-QcCompliance',
              2: 'qcs-QcLimitValue',
              3: 'qcs-QcRetentionPeriod',
              4: 'qcs-QcSSCD',
              5: 'qcs-QcPDS',
              6: {
                null: 'qcs-QcType',
                1: 'qct-esign',
                2: 'qct-eseal',
                3: 'qct-web',
              }
            }
          },
          194121: {
            null: 'qualified-certificate-policies',
            1: {
              null: 'policy-identifiers',
              0: 'qcp-natural',
              1: 'qcp-legal',
              2: 'qcp-natural-qscd',
              3: 'qcp-legal-qscd',
              4: 'qcp-web',
            }
          }
        }
      }
    },
    1: {
      null: 'iso',
      2: {
        null: 'member-body',
        40: {
          10040: {
            null: 'x9-57',
            4: {null: 'x9cm', 1: 'dsa'}
          }
        },
        840: {
          null: 'us',
          10045: {
            null: 'ansi-X9-62',
            2: {
              null: 'keyType',
              1: 'ecPublicKey',
            },
            3: {
              null: 'curves',
              1: {
                null: 'prime',
                2: 'prime192v2',
                3: 'prime192v3',
                4: 'prime239v1',
                5: 'prime239v2',
                6: 'prime239v3',
                7: 'prime256v1',
              }
            }
          },
          113549: {
            null: 'rsadsi',
            1: {
              null: 'pkcs',
              11: 'pkcs-11',
              12: 'pkcs-12',
              1: {
                null: 'pkcs-1',
                1: 'rsaEncryption',
                2: 'md2withRSAEncryption',
                4: 'md5withRSAEncryption',
                5: 'sha1WithRSAEncryption',
                6: 'rsaOAEPEncryptionSET',
                7: 'id-RSAES-OAEP',
                8: 'id-mgf1',
                9: 'id-pSpecified',
                10: 'rsassa-pss',
                11: 'sha256WithRSAEncryption',
                12: 'sha384WithRSAEncryption',
                13: 'sha512WithRSAEncryption',
                14: 'sha224WithRSAEncryption',
                15: 'sha512-224WithRSAEncryption',
                16: 'sha512-256WithRSAEncryption',
              },
              9: {
                null: 'pkcs-9',
                1: 'emailAddress',
                2: 'unstructuredName',
                3: 'contentType',
                4: 'messageDigest',
                5: 'signingTime',
                6: 'countersignature',
                7: 'challengePassword',
                8: 'unstructuredAddress',
                9: 'extendedCertificateAttributes'
              },
              3: {null: 'pkcs-3', 1: 'dhKeyAgreement'},
              7: {
                null: 'pkcs-7',
                1: 'data',
                2: 'signedData',
                3: 'envelopedData',
                4: 'signedAndEnvelopedData',
                5: 'digestData',
                6: 'encryptedData'
              }
            },
            2: {null: 'digestAlgorithm', 2: 'md2', 4: 'md4', 5: 'md5'},
            3: {
              null: 'encryptionAlgorithm',
              2: 'rc2CBC',
              4: 'rc4',
              7: 'DES-EDE3-CBC',
              8: 'RC5CBC',
              9: 'RC5CBCPAD',
              10: 'desCDMF',
              17: 'des-ede3'
            },
            5: {
              null: 'pkcs-5',
              1: 'pbeWithMD2AndDES-CBC',
              3: 'pbeWithMD5AndDES-CBC',
              11: 'pbeWithSHA1AndRC2-CBC',
              12: 'pbeWithSHA1AndRC4'
            },
            12: {
              null: 'pkcs-12',
              1: {
                null: 'pkcs-12PbeIds',
                1: 'pbeWithSHA1And128BitRC4',
                2: 'pbeWithSHA1And40BitRC4',
                3: 'pbeWithSHA1And3-KeyTripleDES-CBC',
                4: 'pbeWithSHA1And2-KeyTripleDES-CBC',
                5: 'pbeWithSHA1And128BitRC2-CBC',
                6: 'pbeWithSHA1And40BitRC2-CBC'
              }
            }
          },
          113533: {
            null: 'nt',
            7: {
              null: 'nsn',
              66: {
                null: 'algorithms',
                10: 'cast5CBC',
                11: 'cast5MAC',
                12: 'pbeWithMD5AndCAST5-CBC'
              }
            }
          }
        }
      },
      3: {
        null: 'identified-organization',
        6: {
          null: 'dod',
          1: {
            null: 'internet',
            4: {
              null: 'private',
              1: {
                null: 'enterprise',
                11129: {
                  null: 'google',
                  2: {
                    null: 'two',
                    4: {
                      null: 'four',
                      2: 'sctList',
                    },
                  },
                },
              },
            },
            5: {
              null: 'security',
              5: {
                7: {
                  null: 'pkix',
                  1: {
                    null: 'pe',
                    1: 'authorityInfoAccess',
                    2: 'biometricInfo',
                    3: 'qcStatements',
                    4: 'auditIdentity',
                    5: 'id-pe-acTargeting',
                    6: 'aaControls',
                    7: {
                      null: 'id-pe-ipAddrBlocks',
                      1: 'id-acTemplate',
                      2: 'id-openPGPCertTemplateExt'
                    },
                    8: 'id-pe-autonomousSysIds',
                    9: 'id-pe-sbgp-routerIdentifier',
                    10: 'proxying',
                    11: 'subjectInfoAccess',
                    12: 'id-pe-logotype',
                    13: 'id-pe-wlanSSID',
                    14: 'id-pe-proxyCertInfo',
                    15: 'id-pe-acPolicies',
                    16: 'id-pe-warranty-extn',
                    17: 'id-pe-sim',
                    18: 'id-pe-cmsContentConstraints',
                    19: 'id-pe-otherCerts',
                    20: 'id-pe-wrappedApexContinKey',
                    21: 'id-pe-clearanceConstraints',
                    22: 'id-pe-skiSemantics',
                    23: 'id-pe-nsa',
                    24: 'ext-TLSFeatures',
                    25: 'id-pe-mud-url',
                    26: 'id-pe-TNAuthList',
                    27: 'id-pe-JWTClaimConstraints',
                    28: 'id-pe-ipAddrBlocks-v2',
                    29: 'id-pe-autonomousSysIds-v2',
                    30: 'id-pe-mudsigner',
                    31: 'id-pe-acmeIdentifier',
                    32: 'id-pe-masa-url',
                  },
                  2: {
                    null: 'qt',
                    1: 'cps',
                    2: 'unotice',
                    3: 'id-qt-textNotice',
                    4: 'id-qt-acps',
                    5: 'id-qt-acunotice'
                  },
                  3: {
                    null: 'kp',
                    1: 'serverAuth',
                    2: 'clientAuth',
                    3: 'codeSigning',
                    4: 'emailProtection',
                    8: 'timeStamping',
                    9: 'OCSPSigning'
                  },
                  11: {
                    null: 'qcs',
                    1: 'pkixQCSyntax-v1',
                    2: 'id-qcs-pkixQCSyntax-v2',
                  }
                }
              }
            }
          }
        },
        14: {
          null: 'oiw',
          3: {
            null: 'secsig',
            2: {
              null: 'algorithm',
              2: 'md4WithRSA',
              3: 'md5WithRSA',
              4: 'md5WithRSAEncryption',
              6: 'desECB',
              7: 'desCBC',
              8: 'desOFB',
              9: 'desCFB',
              10: 'desMAC',
              11: 'RSASignature',
              12: 'DSA',
              13: 'DSAWithSHA',
              14: 'RSAWithmdc2',
              15: 'RSAWithSHA',
              16: 'dhWithCommonModulus',
              17: 'desEDE',
              18: 'SHA',
              19: 'mdc-2',
              20: 'DSACommon',
              21: 'DSACommonWithSHA',
              22: 'RSAKeyTransport',
              23: 'Keyed-hash-seal',
              24: 'md2WithRSASignature',
              25: 'md5WithRSASignature',
              26: 'SHA1',
              27: 'DSAWithSHA1',
              28: 'DSACommonWithSHA1',
              29: 'RSASignatureWithSHA1'
            }
          },
          7: {
            2: {
              1: {1: 'elGamal'},
              3: {1: 'md2WithRsa', 2: 'md2WithElGamal'}
            }
          }
        },
        36: {
          null: 'teletrust',
          3: {
            null: 'algorithm',
            2: {
              null: 'hashAlgorithm',
              1: 'ripemd160',
              2: 'ripemd128',
              3: 'ripemd256'
            },
            3: {
              null: 'signatureAlgorithm',
              1: {
                null: 'rsaSignature',
                2: 'rsaSignatureWithripemd160',
                3: 'rsaSignatureWithripemd128',
                4: 'rsaSignatureWithripemd256'
              }
            }
          }
        },
        132: {
          null: 'certicom',
          0: {
            null: 'curve',
            1: 'sect163k1',
            2: 'sect163r1',
            3: 'sect239k1',
            4: 'sect113r1',
            5: 'sect113r2',
            6: 'secp112r1',
            7: 'secp112r2',
            8: 'secp160r1',
            9: 'secp160k1',
            10: 'secp256k1',
            15: 'sect163r2',
            16: 'sect283k1',
            17: 'sect283r1',
            22: 'sect131r1',
            23: 'sect131r2',
            24: 'sect193r1',
            25: 'sect193r2',
            26: 'sect233k1',
            27: 'sect233r1',
            28: 'secp128r1',
            29: 'secp128r2',
            30: 'secp160r2',
            31: 'secp192k1',
            32: 'secp224k1',
            33: 'secp224r1',
            34: 'secp384r1',
            35: 'secp521r1',
            36: 'sect409k1',
            37: 'sect409r1',
            38: 'sect571k1',
            39: 'sect571r1',
          }
        }
      }
    },
    2: {
      null: 'joint-iso-ccitt',
      5: {
        null: 'ds',
        4: {
          null: 'at',
          3: 'commonName',
          4: 'surname',
          6: 'countryName',
          7: 'localityName',
          8: 'stateOrProvinceName',
          10: 'organizationName',
          11: 'organizationUnitName',
          12: 'title',
          35: 'userPassword',
          36: 'userCertificate',
          37: 'cAcertificate',
          38: 'authorityRecovationList',
          39: 'certificateRevocationList',
          40: 'crossCertificatePair',
          41: 'name',
          42: 'givenName',
          43: 'initials',
          44: 'generationQualifier',
          46: 'dnQualifier',
          58: 'attributeCertificate'
        },
        8: {
          null: 'algorithm',
          1: {null: 'encryptionAlgorithm', 1: 'rsa'},
          2: 'hashAlgorithm',
          3: 'signatureAlgorithm'
        },
        29: {
          null: 'ce',
          9: 'subjectDirectoryAttributes',
          14: 'subjectKeyIdentifier',
          15: 'keyUsage',
          16: 'privateKeyUsagePeriod',
          17: 'subjectAltName',
          18: 'issuerAltName',
          19: 'basicConstraints',
          20: 'cRLNumber',
          21: 'reasonCode',
          23: 'instructionCode',
          24: 'invalidityDate',
          27: 'deltaCRLIndicator',
          28: 'issuingDistributionPoint',
          29: 'certificateIssuer',
          30: 'nameConstraints',
          31: 'cRLDistributionPoints',
          32: 'certificatePolicies',
          33: 'policyMappings',
          35: 'authorityKeyIdentifier',
          36: 'policyConstraints',
          37: 'extKeyUsage'
        }
      },
      16: {
        840: {
          1: {
            113730: {
              null: 'netscape',
              1: {
                null: 'netscape-cert-extension',
                1: 'netscape-cert-extension-type',
                2: 'netscape-base-url',
                3: 'netscape-revocation-url',
                4: 'netscape-ca-revocation-url',
                7: 'netcape-cert-renewal-url',
                8: 'netscape-policy-url',
                12: 'netscape-ssl-server-name',
                13: 'netscape-comment'
              },
              2: {null: 'netscape-data-type', 5: 'netscape-cert-sequence'}
            }
          }
        }
      },
      23: {
        null: 'international-organizations',
        140: {
          null: 'ca-browser-forum',
          1: {
            null: 'certificate-policies',
            1: 'ev-guidelines',
            2: {
              null: 'baseline-requirements',
              1: 'domain-validated',
              2: 'organization-validated',
              3: 'individual-validated'
            },
            3: '3',
            4: {
              null: 'code-signing-requirements',
              1: 'code-signing',
            },
            31: '31'
          }
        }
      }
    }
  };
}

// Throw when There OID name is unknown.
// For example, be defined unique extension.
class UnknownOIDNameError extends StateError {
  UnknownOIDNameError(String message) : super(message);
}
